[compiler] Expose `to_relay` API which hands the Relay graph over in Python

bwasti opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44 [compiler] Expose `to_relay` API which hands the Relay graph over in Python**
* #43 [compiler] Expose relay ops registered as torch.ops.tvm.__

bwasti committed:
[wip] Get relay expr

bwasti committed:
Update on "[wip] Get relay expr"

[compiler] Expose `to_relay` API which hands the Relay graph over in
Python

gh-metadata: pytorch tvm 44 gh/bwasti/28/head

bwasti committed:
Update on "[compiler] Expose `to_relay` API which hands the Relay graph over in Python"

[compiler] Expose `to_relay` API which hands the Relay graph over in
Python

gh-metadata: pytorch tvm 44 gh/bwasti/28/head

wanchaol commented on torch_tvm/register.cpp:
```diff
@@ -72,5 +75,33 @@ PYBIND11_MODULE(_torch_tvm, m) {
 
   m.def("disable", []() { fusion_enabled = false; });
 
+  m.def("push_relay_expr", [](std::shared_ptr<Graph> g,
```
since we don't expose the `push_relay_expr` as a public API, should we make it as `_push_relay_expr`?

wanchaol commented on torch_tvm/register.cpp:
```diff
@@ -21,11 +23,12 @@ static int opt_level = 2;
+static auto tvm_sym = Symbol::fromQualString("tvm::CompilationGroup");
 
+static std::unordered_map<size_t, tvm::relay::Expr> relay_exprs;
+static size_t relay_exprs_uuid = 1337;
```
what is this number specially for? I think it can also starts at 0?

wanchaol commented on torch_tvm/compiler.cpp:
```diff
@@ -27,11 +27,11 @@ tvm::relay::Var TVMCompiler::convertToRelay(Value* val) {
   // All doubles are converted to floats
   if (val.isDouble()) {
     auto x = tvm::runtime::NDArray::Empty(
-        {}, tvm::runtime::String2TVMType("float32"), ctx_);
```
I am a bit surprised this could access the`ctx_` before..

bwasti replied:
it's a trick! this wasn't a static method before

wanchaol commented on torch_tvm/register.cpp:
```diff
@@ -72,5 +75,33 @@ PYBIND11_MODULE(_torch_tvm, m) {
+        ctx.device_id = 0;
+        auto expr = TVMCompiler::convertToRelay(subgraph, ctx);
+        relay_exprs[++relay_exprs_uuid] = expr;
+        return relay_exprs_uuid;
```
that will only convert the first subgraph to relay right? Shall we give the user warning here indicating they could not convert the full graph to relay, they will need to add more operators to make sure a single subgraph

wanchaol commented on torch_tvm/register.cpp:
```diff
@@ -1,9 +1,11 @@
 #include <pybind11/pybind11.h>
+#include <stack>
```
I did not see where this header is needed?

wanchaol reviewed (approved):
looks nice! some nits

bwasti committed:
Update on "[compiler] Expose `to_relay` API which hands the Relay graph over in Python"

[compiler] Expose `to_relay` API which hands the Relay graph over in
Python

gh-metadata: pytorch tvm 44 gh/bwasti/28/head

bwasti committed:
Update on "[compiler] Expose `to_relay` API which hands the Relay graph over in Python"

[compiler] Expose `to_relay` API which hands the Relay graph over in
Python

gh-metadata: pytorch tvm 44 gh/bwasti/28/head
